File: math-power.cpp
/**
 * (c) 2024, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */

#include "pxt.h"
#include <cmath>


namespace Math {

    //%
    TNumber power(TNumber x, TNumber y) {
		double dx = toDouble(x);
		double dy = toDouble(y);
		if ( ::floor( dy) == dy)
		{
			return fromDouble(__builtin_powi(dx, dy));
		}
		if ( dx > 0)
		{
			return fromDouble( ::exp( dy * ::log( dx)));
		}
		return fromDouble( ::log( dx));
    }	
}

File: math-power.ts
/**
 * (c) 2024, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */

namespace Math {
    /**
     * Returns the value of a base expression taken to a specified power.
     * @param x The base value of the expression.
     * @param y The exponent value of the expression.
     */
	//% shim=Math::power
    //% block="$x|to power $y"
    //% x.defl=1
    //% y.defl=1
    export function power(x: number, y: number): number {
        if (Math.floor(y) == y) {
			Math.pow(x, y)
		}
        if (x > 0) {
            return Math.exp(y * Math.log(x))
        }
        return Math.log(x)
    }
}
